torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -160,8 +161,21 @@ void initializeStreamsEvents(
 
     // `tensors` are created on a different stream. Hence, they must record
     // new streams in this Work to prevent being freed before the Work finishes.
-    c10::cuda::CUDACachingAllocator::recordStream(
-      tensors[i].storage().data(), streams[i]);
+    if (tensors[i].is_sparse()) {
+      if (tensors[i].is_coalesced()) {
+        c10::cuda::CUDACachingAllocator::recordStream(
+            tensors[i].indices().storage().data(), streams[i]);
+        c10::cuda::CUDACachingAllocator::recordStream(
+            tensors[i].values().storage().data(), streams[i]);
+      } else {
+        // We will need to coalesce first, which means new tensors will
+        // be allocated on the streams we just allocated, and there
+        // is no need to record them separately.
```

Comment:
It won't be freed because the op holds on to the input tensors. The coalesce on this new stream will happen after any operations against the inputs because of `events[i].block()` that is run before this block.